runtime/src/main/java/org/corfudb/runtime/collections/TxnContextShim.java

```diff
@@ -175,196 +227,49 @@ TxnContextShim putRecord(@Nonnull Table<K, V, M> table,
     }
 
     /**
-     * touch() a key to generate a conflict on it. The value will not be modified.
-     *
-     * @param table    Table object to perform the touch() in.
-     * @param key      Key of the record.
-     * @param <K>      Type of Key.
-     * @param <V>      Type of Value.
-     * @param <M>      Type of Metadata.
-     * @return TxnContext instance.
-     * @throws UnsupportedOperationException if attempted on a non-existing object.
-     */
-    @Nonnull
-    public <K extends Message, V extends Message, M extends Message>
-    TxnContextShim touch(@Nonnull Table<K, V, M> table,
-                         @Nonnull final K key) {
-        this.txnContext = txnContext.touch(table, key);
-        return this;
-    }
-
-    /**
-     * touch() variant that accepts a table name instead of table object.
-     */
-    @Nonnull
-    public <K extends Message, V extends Message, M extends Message>
-    TxnContextShim touch(@Nonnull String tableName,
-                         @Nonnull final K key) {
-        return this.touch(txnContext.getTable(tableName), key);
-    }
-
-    /**
-     *
-     * @param table - table object to test if record exists
-     * @param key - key or identifier to test for existence.
-     * @param <K> - type of the key
-     * @param <V> - type of payload or value
-     * @param <M> - type of metadata
-     * @return true if record exists and false if record does not exist.
+     * Protobuf objects are immutable. So any metadata modifications made here won't be
+     * reflected back into the caller's in-memory object directly.
+     * The caller is only really interested in the modified values of those transactions
+     * that successfully commit.
+     * To reflect metadata changes made here, we modify commit() to accept a callback
+     * that carries all the final values of the changes made by this transaction.
      */
-    @Nonnull
-    public <K extends Message, V extends Message, M extends Message>
-    boolean isExists(@Nonnull Table<K, V, M> table, @Nonnull final K key) {
-        CorfuStoreEntry<K, V, M> record = this.txnContext.getRecord(table, key);
-        return record.getPayload() != null;
+    public interface CommitCallback {
+        /**
+         * This callback returns a list of stream entries as opposed to CorfuStoreEntries
+         * because if this transaction had operations like clear() then the CorfuStoreEntry
+         * would just be empty.
+         *
+         * @param mutations - A group of all tables touched by this transaction along with
+         *                    the updates made in each table.
+         */
+        void onCommit(Map<String, List<CorfuStreamEntry>> mutations);
```

Comment:
wont the applications need significant corfu-internal understanding to process these entries?

Reply:
This is the same callback we have for stream entry processing, it really just has key, value and metadata, so no corfu internals are exposed.